### PR TITLE
add back FI_EFA_USE_DEVICE_RDMA=1 flag, revert 2936f22

### PR DESCRIPTION
**Pull request by ydaiming**
*Issue #, if available:*

Without the `FI_EFA_USE_DEVICE_RDMA=1`, the NCCL performance is many times slower than with the flag. The flag was removed in https://github.com/aws/sagemaker-training-toolkit/pull/106 because of incompatibility, but should be re-enabled with latest `EFA` with updated `libfabric` and newer `OFI` packages. This PR also updates and fixes `black` lint issues.

*Description of changes:*

Add back `FI_EFA_USE_DEVICE_RDMA=1` flag for faster NCCL performance in `smdataparallel` launcher and unit tests.

feature: add back FI_EFA_USE_DEVICE_RDMA=1 flag, revert 2936f22 
fix: fixed the `black` lint, upgraded black to version 21.3.0
fix: remove u prefix of strings, as python3 defaults to unicode strings

note: EFA is only available on p3dn or p4dn instances
note: EFA version 1.15.1 and OFI 1.1.5-aws have the issue fixed
note: black format reference on remove u prefix
https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#strings

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-training-toolkit/blob/master/CONTRIBUTING.md) doc
- [ ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-training-toolkit/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have used the regional endpoint when creating S3 and/or STS clients (if appropriate)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-training-toolkit/blob/master/README.md)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by josephevans — DISMISSED** [review]

**Comment by satishpasumarthi** [comment]
LGTM !

**Comment by rashika** [comment]
Is there a related patch which confirms that the latest EFA installer is being used since this flag has dependency on that.

File: src/sagemaker_training/smdataparallel.py [hunk 1]
```diff
@@ -164,6 +164,8 @@ def _get_mpirun_command(
             "RDMAV_FORK_SAFE=1",
             "-x",
             "LD_PRELOAD=%s" % inspect.getfile(gethostname),
+            "-x",
```
Comment (satishpasumarthi) [inline src/sagemaker_training/smdataparallel.py:167]: EFA is enabled on only certain instance types. Please add a check and use extend to add this option.

**Review by satishpasumarthi — CHANGES_REQUESTED** [review]
Left a comment

File: src/sagemaker_training/smdataparallel.py (hunk 1 above)
Comment (ydaiming) [reply src/sagemaker_training/smdataparallel.py]: EFA is only enabled on `["ml.p3dn.24xlarge", "ml.p4d.24xlarge"]`. added the check in the commit now.

**Comment by ydaiming** [comment]
@rashika, We have a check on the instance type for this flag.
There's no specific check for EFA. The toolkit is mostly run in controlled environments (DLCs), and checking specifically on libfabric, OFI, or EFA isn't easy. It's not in the scope of this PR.

**Comment by satishpasumarthi** [comment]
LGTM !

**Review by satishpasumarthi — APPROVED** [review]
LGTM !

**Pull request merged**